vespalib/src/vespa/vespalib/portal/reactor.cpp

```diff
@@ -0,0 +1,128 @@
+// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+
+#include "reactor.h"
+#include <cassert>
+
+namespace vespalib::portal {
+
+Reactor::EventHandler::~EventHandler() = default;
+
+//-----------------------------------------------------------------------------
+
+Reactor::Token::Token(Reactor &reactor, EventHandler &handler, int fd, bool read, bool write)
+    : _reactor(reactor), _handler(handler), _fd(fd)
+{
+    ++_reactor._token_cnt;
+    _reactor._selector.add(_fd, _handler, read, write);
+}
+
+void
+Reactor::Token::update(bool read, bool write)
+{
+    _reactor._selector.update(_fd, _handler, read, write);
+}
+
+Reactor::Token::~Token()
+{
+    _reactor._selector.remove(_fd);
+    _reactor.cancel_token(*this);
+    --_reactor._token_cnt;
+}
+
+//-----------------------------------------------------------------------------
+
+void
+Reactor::cancel_token(const Token &)
+{
+    if (std::this_thread::get_id() == _thread.get_id()) {
+        _skip_events = true;
```

Comment:
To make sure I'm following the semantics correctly, I'm assuming this is for dealing with tokens cancelled recursively as part of event handlers being invoked?

Reply:
Yes, this is to ensure that tokens cancelled during a callback from the reactor thread will not receive any additional events. Due to epoll event collapsing you will only extract a single event for each event handler each time epoll wait is called, so self cancellation is not really an issue. This is to avoid delivering events that have been extracted, but then cancelled by the handling of an earlier event for a different handler. It is a very simple event filtering mechanism. A more fleshed out alternative would be to keep a list/map of event handlers that should not be invoked.